<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Comparison of different nucleotide or amino acid substitution models"><title>ModelTest — modelTest • phangorn</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="ModelTest — modelTest"><meta property="og:description" content="Comparison of different nucleotide or amino acid substitution models"><meta property="og:image" content="https://klausvigo.github.io/phangorn/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">phangorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.10.0.900</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Trees.html">Estimating phylogenetic trees with phangorn</a>
    <a class="dropdown-item" href="../articles/Morphological.html">Phylogenetic trees from morphological data</a>
    <a class="dropdown-item" href="../articles/AdvancedFeatures.html">Markov models and transition rate matrices</a>
    <a class="dropdown-item" href="../articles/Ancestral.html">Ancestral Sequence Reconstruction</a>
    <a class="dropdown-item" href="../articles/Networx.html">Splits and Networx</a>
    <a class="dropdown-item" href="../articles/MLbyHand.html">Maximum likelihood by hand</a>
    <a class="dropdown-item" href="../articles/IntertwiningTreesAndNetworks.html">Intertwining phylogenetic trees and networks: R Example Script</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/KlausVigo/phangorn/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>ModelTest</h1>
      <small class="dont-index">Source: <a href="https://github.com/KlausVigo/phangorn/blob/HEAD/R/modelTest.R" class="external-link"><code>R/modelTest.R</code></a></small>
      <div class="d-none name"><code>modelTest.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Comparison of different nucleotide or amino acid substitution models</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">modelTest</span><span class="op">(</span><span class="va">object</span>, tree <span class="op">=</span> <span class="cn">NULL</span>, model <span class="op">=</span> <span class="cn">NULL</span>, G <span class="op">=</span> <span class="cn">TRUE</span>, I <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  FREQ <span class="op">=</span> <span class="cn">FALSE</span>, k <span class="op">=</span> <span class="fl">4</span>, control <span class="op">=</span> <span class="fu"><a href="pml.control.html">pml.control</a></span><span class="op">(</span>epsilon <span class="op">=</span> <span class="fl">1e-08</span>, maxit <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  trace <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, multicore <span class="op">=</span> <span class="cn">FALSE</span>, mc.cores <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p>an object of class phyDat or pml</p></dd>


<dt>tree</dt>
<dd><p>a phylogenetic tree.</p></dd>


<dt>model</dt>
<dd><p>a vector containing the substitution models to compare with
each other or "all" to test all available models</p></dd>


<dt>G</dt>
<dd><p>logical, TRUE (default) if (discrete) Gamma model should be tested</p></dd>


<dt>I</dt>
<dd><p>logical, TRUE (default) if invariant sites should be tested</p></dd>


<dt>FREQ</dt>
<dd><p>logical, FALSE (default) if TRUE amino acid frequencies will be
estimated.</p></dd>


<dt>k</dt>
<dd><p>number of rate classes</p></dd>


<dt>control</dt>
<dd><p>A list of parameters for controlling the fitting process.</p></dd>


<dt>multicore</dt>
<dd><p>logical, whether models should estimated in parallel.</p></dd>


<dt>mc.cores</dt>
<dd><p>The number of cores to use, i.e. at most how many child
processes will be run simultaneously. Must be at least one, and
parallelization requires at least two cores.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A data.frame containing the log-likelihood, number of estimated
parameters, AIC, AICc and BIC all tested models.  The data.frame has an
attributes "env" which is an environment which contains all the trees, the
data and the calls to allow get the estimated models, e.g. as a starting
point for further analysis (see example).</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><code>modelTest</code> estimates all the specified models for a given tree and
data.  When the mclapply is available, the computations are done in
parallel. <code>modelTest</code> runs each model in one thread.  This is may not
work within a GUI interface and will not work under Windows.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Burnham, K. P. and Anderson, D. R (2002) <em>Model selection
and multimodel inference: a practical information-theoretic approach</em>. 2nd
ed. Springer, New York</p>
<p>Posada, D. and Crandall, K.A. (1998) MODELTEST: testing the model of DNA
substitution. <em>Bioinformatics</em> <b>14(9)</b>: 817-818</p>
<p>Posada, D. (2008) jModelTest: Phylogenetic Model Averaging. <em>Molecular
Biology and Evolution</em> <b>25</b>: 1253-1256</p>
<p>Darriba D., Taboada G.L., Doallo R and Posada D. (2011) ProtTest 3: fast
selection of best-fit models of protein evolution. . <em>Bioinformatics</em>
<b>27</b>: 1164-1165</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="pml.html">pml</a></code>, <code><a href="https://rdrr.io/r/stats/anova.html" class="external-link">anova</a></code>, <code><a href="https://rdrr.io/r/stats/AIC.html" class="external-link">AIC</a></code>,
<code><a href="codonTest.html">codonTest</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Klaus Schliep <a href="mailto:klaus.schliep@gmail.com">klaus.schliep@gmail.com</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/example.html" class="external-link">example</a></span><span class="op">(</span><span class="va">NJ</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">(</span><span class="va">mT</span> <span class="op">&lt;-</span> <span class="fu">modelTest</span><span class="op">(</span><span class="va">Laurasiatherian</span>, <span class="va">tree</span>, model <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"JC"</span>, <span class="st">"F81"</span>, <span class="st">"K80"</span>, <span class="st">"HKY"</span>,</span></span>
<span class="r-in"><span>                 <span class="st">"SYM"</span>, <span class="st">"GTR"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># extract best model</span></span></span>
<span class="r-in"><span><span class="op">(</span><span class="va">best_model</span> <span class="op">&lt;-</span> <span class="fu"><a href="pml.html">as.pml</a></span><span class="op">(</span><span class="va">mT</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">chloroplast</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">(</span><span class="va">mTAA</span> <span class="op">&lt;-</span> <span class="fu">modelTest</span><span class="op">(</span><span class="va">chloroplast</span>, model<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"JTT"</span>, <span class="st">"WAG"</span>, <span class="st">"LG"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># test all available amino acid models</span></span></span>
<span class="r-in"><span><span class="op">(</span><span class="va">mTAA_all</span> <span class="op">&lt;-</span> <span class="fu">modelTest</span><span class="op">(</span><span class="va">chloroplast</span>, model<span class="op">=</span><span class="st">"all"</span>, multicore<span class="op">=</span><span class="cn">TRUE</span>, mc.cores<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Klaus Schliep, Emmanuel Paradis, Leonardo de Oliveira Martins, Alastair Potts, Iris Bardel-Kahr.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

