<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="phangorn">
<title>Ancestral Sequence Reconstruction • phangorn</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Ancestral Sequence Reconstruction">
<meta property="og:description" content="phangorn">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">phangorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.8.1.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/AdvancedFeatures.html">Advanced features</a>
    <a class="dropdown-item" href="../articles/Ancestral.html">Ancestral Sequence Reconstruction</a>
    <a class="dropdown-item" href="../articles/IntertwiningTreesAndNetworks.html">Intertwining phylogenetic trees and networks: R Example Script</a>
    <a class="dropdown-item" href="../articles/Networx.html">Splits and Networx</a>
    <a class="dropdown-item" href="../articles/Trees.html">Estimating phylogenetic trees with phangorn</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/KlausVigo/phangorn/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Ancestral Sequence Reconstruction</h1>
                        <h4 data-toc-skip class="author">Klaus Schliep</h4>
            <address class="author_afil">
      Graz University of Technology<br><a class="author_email" href="mailto:#"></a><a href="mailto:klaus.schliep@gmail.com" class="email">klaus.schliep@gmail.com</a>
      </address>
                  
            <h4 data-toc-skip class="date">2022-06-07</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/KlausVigo/phangorn/blob/HEAD/vignettes/Ancestral.Rmd" class="external-link"><code>vignettes/Ancestral.Rmd</code></a></small>
      <div class="d-none name"><code>Ancestral.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>These notes describe the ancestral sequence reconstruction using the <em>phangorn</em> package <span class="citation">(Schliep 2011)</span>. <em>phangorn</em> provides several methods to estimate ancestral character states with either Maximum Parsimony (MP) or Maximum Likelihood (ML). For more background on all the methods see e.g. <span class="citation">(Felsenstein 2004)</span> or <span class="citation">(Yang 2006)</span>.</p>
</div>
<div class="section level2">
<h2 id="parsimony-reconstructions">Parsimony reconstructions<a class="anchor" aria-label="anchor" href="#parsimony-reconstructions"></a>
</h2>
<p>To reconstruct ancestral sequences we first load some data and reconstruct a tree:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/KlausVigo/phangorn" class="external-link">phangorn</a></span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Loading required package: ape</span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/trees"</span>, package <span class="op">=</span> <span class="st">"phangorn"</span><span class="op">)</span>
<span class="va">primates</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read.phyDat.html">read.phyDat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>, <span class="st">"primates.dna"</span><span class="op">)</span>,
                        format <span class="op">=</span> <span class="st">"interleaved"</span><span class="op">)</span>
<span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parsimony.html">pratchet</a></span><span class="op">(</span><span class="va">primates</span>, trace<span class="op">=</span><span class="fl">0</span><span class="op">)</span> |&gt; <span class="fu"><a href="../reference/parsimony.html">acctran</a></span><span class="op">(</span><span class="va">primates</span><span class="op">)</span>
<span class="fu"><a href="../reference/parsimony.html">parsimony</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">primates</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 746</span></code></pre>
<p>For parsimony analysis of the edge length represent the observed number of changes. Reconstructing ancestral states therefore defines also the edge lengths of a tree. However there can exist several equally parsimonious reconstructions or states can be ambiguous and therefore edge length can differ. “MPR” reconstructs the ancestral states for each (internal) node as if the tree would be rooted in that node. However the nodes are not independent of each other. If one chooses one state for a specific node, this can restrict the choice of neighboring nodes (figures 2 and 3). The function acctran (accelerated transformation) assigns edge length and internal nodes to the tree <span class="citation">(Swofford and Maddison 1987)</span>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">anc.acctran</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ancestral.pml.html">ancestral.pars</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">primates</span>, <span class="st">"ACCTRAN"</span><span class="op">)</span>
<span class="va">anc.mpr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ancestral.pml.html">ancestral.pars</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">primates</span>, <span class="st">"MPR"</span><span class="op">)</span></code></pre></div>
<p>All the ancestral reconstructions for parsimony are based on the fitch algorithm and so far only bifurcating trees are allowed. However trees can get pruned afterwards using the function <code>multi2di</code> from <em>ape</em>.</p>
<p>The <code>seqLogo</code> function from the <em>seqLogo</em> package from Bioconductor provides a neat way to show proportions of a nucleotides of ancestral states (see figure 1).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">seqLogo</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/seqLogo/man/seqLogo.html" class="external-link">seqLogo</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">anc.mpr</span>, <span class="fu"><a href="../reference/midpoint.html">getRoot</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, ic.scale<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p><img src="seqLogo.png"></p>
<p>You may need to install <em>seqLogo</em> before</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"seqLogo"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/ancestral.pml.html">plotAnc</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">anc.mpr</span>, <span class="fl">17</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/title.html" class="external-link">title</a></span><span class="op">(</span><span class="st">"MPR"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="Ancestral_files/figure-html/MPR-1.png" alt="Fig 2. Ancestral reconstruction using MPR." width="576"><p class="caption">
Fig 2. Ancestral reconstruction using MPR.
</p>
</div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/ancestral.pml.html">plotAnc</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">anc.acctran</span>, <span class="fl">17</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/title.html" class="external-link">title</a></span><span class="op">(</span><span class="st">"ACCTRAN"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="Ancestral_files/figure-html/ACCTRAN-1.png" alt="Fig 3. Ancestral reconstruction using ACCTRAN." width="576"><p class="caption">
Fig 3. Ancestral reconstruction using ACCTRAN.
</p>
</div>
</div>
<div class="section level2">
<h2 id="likelihood-reconstructions">Likelihood reconstructions<a class="anchor" aria-label="anchor" href="#likelihood-reconstructions"></a>
</h2>
<p><em>phangorn</em> also offers the possibility to estimate ancestral states using a ML. The advantages of ML over parsimony is that the reconstruction accounts for different edge lengths. So far only a marginal construction is implemented (see <span class="citation">(Yang 2006)</span>).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">pml</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">primates</span><span class="op">)</span>
<span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">optim.pml</a></span><span class="op">(</span><span class="va">fit</span>, model<span class="op">=</span><span class="st">"F81"</span>, control <span class="op">=</span> <span class="fu"><a href="../reference/pml.control.html">pml.control</a></span><span class="op">(</span>trace<span class="op">=</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>We can assign the ancestral states according to the highest likelihood (“ml”): <span class="math display">\[
P(x_r = A) =  \frac{L(x_r=A)}{\sum_{k \in \{A,C,G,T\}}L(x_r=k)}
\]</span> and the highest posterior probability (“bayes”) criterion: <span class="math display">\[
P(x_r=A) =  \frac{\pi_A L(x_r=A)}{\sum_{k \in \{A,C,G,T\}}\pi_k L(x_r=k)},
\]</span> where <span class="math inline">\(L(x_r)\)</span> is the joint probability of states at the tips and the state at the root <span class="math inline">\(x_r\)</span> and <span class="math inline">\(\pi_i\)</span> are the estimated base frequencies of state <span class="math inline">\(i\)</span>. Both methods agree if all states (base frequencies) have equal probabilities.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">anc.ml</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ancestral.pml.html">ancestral.pml</a></span><span class="op">(</span><span class="va">fit</span>, <span class="st">"ml"</span><span class="op">)</span>
<span class="va">anc.bayes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ancestral.pml.html">ancestral.pml</a></span><span class="op">(</span><span class="va">fit</span>, <span class="st">"bayes"</span><span class="op">)</span></code></pre></div>
<p>The differences of the two approaches for a specific site (17) are represented in the following figures.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/ancestral.pml.html">plotAnc</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">anc.ml</span>, <span class="fl">17</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/title.html" class="external-link">title</a></span><span class="op">(</span><span class="st">"ML"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="Ancestral_files/figure-html/plotML-1.png" alt="Fig 4. Ancestral reconstruction the using the maximum likelihood." width="576"><p class="caption">
Fig 4. Ancestral reconstruction the using the maximum likelihood.
</p>
</div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/ancestral.pml.html">plotAnc</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">anc.bayes</span>, <span class="fl">17</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/title.html" class="external-link">title</a></span><span class="op">(</span><span class="st">"Bayes"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="Ancestral_files/figure-html/plotB-1.png" alt="Fig 5. Ancestral reconstruction using (empirical) Bayes." width="576"><p class="caption">
Fig 5. Ancestral reconstruction using (empirical) Bayes.
</p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span class="co">## R version 4.2.0 (2022-04-22)</span>
<span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">## Running under: Ubuntu 20.04.4 LTS</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0</span>
<span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span>
<span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span>
<span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span>
<span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">## [1] phangorn_2.8.1.1 ape_5.6-2       </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##  [1] Rcpp_1.0.8.3      highr_0.9         bslib_0.3.1       compiler_4.2.0   </span>
<span class="co">##  [5] jquerylib_0.1.4   prettydoc_0.4.1   tools_4.2.0       digest_0.6.29    </span>
<span class="co">##  [9] jsonlite_1.8.0    evaluate_0.15     memoise_2.0.1     nlme_3.1-157     </span>
<span class="co">## [13] lattice_0.20-45   pkgconfig_2.0.3   rlang_1.0.2       Matrix_1.4-1     </span>
<span class="co">## [17] fastmatch_1.1-3   igraph_1.3.1      cli_3.3.0         yaml_2.3.5       </span>
<span class="co">## [21] parallel_4.2.0    pkgdown_2.0.3     xfun_0.31         fastmap_1.1.0    </span>
<span class="co">## [25] stringr_1.4.0     knitr_1.39        desc_1.4.1        generics_0.1.2   </span>
<span class="co">## [29] fs_1.5.2          sass_0.4.1        systemfonts_1.0.4 rprojroot_2.0.3  </span>
<span class="co">## [33] grid_4.2.0        R6_2.5.1          textshaping_0.3.6 rmarkdown_2.14   </span>
<span class="co">## [37] purrr_0.3.4       magrittr_2.0.3    codetools_0.2-18  htmltools_0.5.2  </span>
<span class="co">## [41] quadprog_1.5-8    ragg_1.2.2        stringi_1.7.6     cachem_1.0.6</span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Felsenstein2004" class="csl-entry">
Felsenstein, Joseph. 2004. <em>Inferring Phylogenies</em>. Sunderland: Sinauer Associates.
</div>
<div id="ref-Schliep2011" class="csl-entry">
Schliep, Klaus Peter. 2011. <span>“Phangorn: Phylogenetic Analysis in <span>R</span>.”</span> <em>Bioinformatics</em> 27 (4): 592–93. <a href="https://doi.org/10.1093/bioinformatics/btq706" class="external-link">https://doi.org/10.1093/bioinformatics/btq706</a>.
</div>
<div id="ref-Swofford1987" class="csl-entry">
Swofford, D. L., and W. P. Maddison. 1987. <span>“Reconstructing Ancestral Character States Under Wagner Parsimony.”</span> <em>Math. Biosci.</em> 87: 199–229.
</div>
<div id="ref-Yang2006" class="csl-entry">
Yang, Ziheng. 2006. <em>Computational Molecular Evolution</em>. Oxford: Oxford University Press.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Klaus Schliep, Emmanuel Paradis, Leonardo de Oliveira Martins, Alastair Potts, Tim W. White.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
