<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="phangorn">
<title>Estimating phylogenetic trees with phangorn • phangorn</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Estimating phylogenetic trees with phangorn">
<meta property="og:description" content="phangorn">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">phangorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.9.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/AdvancedFeatures.html">Advanced features</a>
    <a class="dropdown-item" href="../articles/Ancestral.html">Ancestral Sequence Reconstruction</a>
    <a class="dropdown-item" href="../articles/IntertwiningTreesAndNetworks.html">Intertwining phylogenetic trees and networks: R Example Script</a>
    <a class="dropdown-item" href="../articles/Networx.html">Splits and Networx</a>
    <a class="dropdown-item" href="../articles/Trees.html">Estimating phylogenetic trees with phangorn</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/KlausVigo/phangorn/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Estimating phylogenetic trees with phangorn</h1>
                        <h4 data-toc-skip class="author">Klaus Schliep</h4>
            <address class="author_afil">
      Graz University of Technology<br><a class="author_email" href="mailto:#"></a><a href="mailto:klaus.schliep@gmail.com" class="email">klaus.schliep@gmail.com</a>
      </address>
                  
            <h4 data-toc-skip class="date">2022-06-23</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/KlausVigo/phangorn/blob/HEAD/vignettes/Trees.Rmd" class="external-link"><code>vignettes/Trees.Rmd</code></a></small>
      <div class="d-none name"><code>Trees.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>These notes should enable the user to estimate phylogenetic trees from alignment data with different methods using the <em>phangorn</em> package <span class="citation">(Schliep 2011)</span> . Several functions of <em>package</em> are also described in more detail in <span class="citation">(Paradis 2012)</span>. For more theoretical background on all the methods see e.g. <span class="citation">(Felsenstein 2004)</span><span class="citation">(Yang 2006)</span>. This document illustrates some of the <code>package</code> features to estimate phylogenetic trees using different reconstruction methods. Small adaptations to the scripts in section @ref(appendix) should enable the user to perform phylogenetic analyses.</p>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>The first thing we have to do is to read in an alignment. Unfortunately there exists many different file formats that alignments can be stored in. The function <code>read.phyDat</code> is used to read in an alignment. There are several functions to read in alignments depending on the format of the data set (“nexus”, “phylip”, “fasta”) and the kind of data (amino acid or nucleotides) in the <em>ape</em> package <span class="citation">(Paradis and Schliep 2019)</span> and <em>phangorn</em>. The function <code>read.phyDat</code> calls these other functions and transform them into a <code>phyDat</code> object. For the specific parameter settings available look in the help files of the function <code>read.dna</code> (for phylip, fasta, clustal format), <code>read.nexus.data</code> for nexus files. For amino acid data additional <code>read.aa</code> is called.</p>
<p>We start our analysis loading the <em>phangorn</em> package and then reading in an alignment.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://ape-package.ird.fr/" class="external-link">ape</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/KlausVigo/phangorn" class="external-link">phangorn</a></span><span class="op">)</span>
<span class="va">fdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/trees"</span>, package <span class="op">=</span> <span class="st">"phangorn"</span><span class="op">)</span>
<span class="va">primates</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read.phyDat.html">read.phyDat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>, <span class="st">"primates.dna"</span><span class="op">)</span>,
                        format <span class="op">=</span> <span class="st">"interleaved"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="distance-based-methods">Distance based methods<a class="anchor" aria-label="anchor" href="#distance-based-methods"></a>
</h2>
<p>After reading in the alignment we can build a first tree with distance based methods. The function dist.dna from the ape package computes distances for many DNA substitution models. To use the function dist.dna we have to transform the data to class DNAbin. For amino acids the function <code>dist.ml</code> offers common substitution models (for example “WAG”, “JTT”, “LG”, “Dayhoff”, “cpREV”, “mtmam”, “mtArt”, “MtZoa” or “mtREV24”).</p>
<p>After constructing a distance matrix we reconstruct a rooted tree with UPGMA and alternatively an unrooted tree using Neighbor Joining <span class="citation">(Saitou and Nei 1987)</span><span class="citation">(Studier and Keppler 1988)</span>. More distance methods like <code>fastme</code> are available in the <em>ape</em> package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dm</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dist.hamming.html">dist.ml</a></span><span class="op">(</span><span class="va">primates</span><span class="op">)</span>
<span class="va">treeUPGMA</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/upgma.html">upgma</a></span><span class="op">(</span><span class="va">dm</span><span class="op">)</span>
<span class="va">treeNJ</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/NJ.html">NJ</a></span><span class="op">(</span><span class="va">dm</span><span class="op">)</span></code></pre></div>
<p>We can plot the trees treeUPGMA and treeNJ with the commands:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">treeUPGMA</span>, main<span class="op">=</span><span class="st">"UPGMA"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="Trees_files/figure-html/plot1-1.png" alt="Rooted UPGMA tree." width="700"><p class="caption">
Rooted UPGMA tree.
</p>
</div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">treeNJ</span>, <span class="st">"unrooted"</span>, main<span class="op">=</span><span class="st">"NJ"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="Trees_files/figure-html/plot2-1.png" alt="Unrooted NJ tree." width="700"><p class="caption">
Unrooted NJ tree.
</p>
</div>
<div class="section level3">
<h3 id="bootstrap">Bootstrap<a class="anchor" aria-label="anchor" href="#bootstrap"></a>
</h3>
<p>To run the bootstrap we need to first write a function which computes a tree from an alignment. So we first need to compute a distance matrix and afterwards compute the tree. This function we can than give to the <code>bootstrap.phyDat</code> function.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fun</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="../reference/upgma.html">upgma</a></span><span class="op">(</span><span class="fu"><a href="../reference/dist.hamming.html">dist.ml</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span>
<span class="va">bs_upgma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootstrap.pml.html">bootstrap.phyDat</a></span><span class="op">(</span><span class="va">primates</span>,  <span class="va">fun</span><span class="op">)</span></code></pre></div>
<p>With the new syntax of R 4.1 this can be written a bit shorter:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>bs_upgma <span class="ot">&lt;-</span> <span class="fu">bootstrap.phyDat</span>(primates,  \(x){<span class="fu">dist.ml</span>(x) <span class="sc">|&gt;</span> upgma})</span></code></pre></div>
<p>Finally we can plot the tree with bootstrap values added:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotBS.html">plotBS</a></span><span class="op">(</span><span class="va">treeUPGMA</span>, <span class="va">bs_upgma</span>, main<span class="op">=</span><span class="st">"UPGMA"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="Trees_files/figure-html/plot_bs-1.png" alt="Rooted UPGMA tree." width="700"><p class="caption">
Rooted UPGMA tree.
</p>
</div>
<p>Distance based methods are very fast and we will use the UPGMA and NJ tree as starting trees for the maximum parsimony and maximum likelihood analyses.</p>
</div>
</div>
<div class="section level2">
<h2 id="parsimony">Parsimony<a class="anchor" aria-label="anchor" href="#parsimony"></a>
</h2>
<p>The function parsimony returns the parsimony score, that is the number of changes which are at least necessary to describe the data for a given tree. We can compare the parsimony score or the two trees we computed so far:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/parsimony.html">parsimony</a></span><span class="op">(</span><span class="va">treeUPGMA</span>, <span class="va">primates</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 751</span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/parsimony.html">parsimony</a></span><span class="op">(</span><span class="va">treeNJ</span>, <span class="va">primates</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 746</span></code></pre>
<p>The function <code>optim.parsimony</code> performs tree rearrangements to find trees with a lower parsimony score. The tree rearrangement implemented are nearest-neighbor interchanges (NNI) and subtree pruning and regrafting (SPR). The later one only works so far with the fitch algorithm.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">treePars</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parsimony.html">optim.parsimony</a></span><span class="op">(</span><span class="va">treeUPGMA</span>, <span class="va">primates</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Final p-score 746 after  1 nni operations</span></code></pre>
<p>However is also a version of the parsimony ratchet <span class="citation">(Nixon 1999)</span> implemented, which is likely to find better trees than just doing NNI / SPR rearrangements. The current implementation</p>
<ol style="list-style-type: decimal">
<li>Create a bootstrap data set <span class="math inline">\(D_b\)</span> from the original data set.</li>
<li>Take the current best tree and perform tree rearrangements on <span class="math inline">\(D_b\)</span> and save bootstrap tree as <span class="math inline">\(T_b\)</span>.</li>
<li>Use <span class="math inline">\(T_b\)</span> and perform tree rearrangements on the original data set. If this trees has a lower parsimony score than the currently best tree replace it.</li>
<li>Iterate 1:3 until either a given number of iteration is reached or no improvements have been recorded for a number of iterations.</li>
</ol>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">treeRatchet</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parsimony.html">pratchet</a></span><span class="op">(</span><span class="va">primates</span>, trace <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>
<span class="fu"><a href="../reference/parsimony.html">parsimony</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">treePars</span>, <span class="va">treeRatchet</span><span class="op">)</span>, <span class="va">primates</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 746 746</span></code></pre>
<p>Next we assign branch length to the tree. The branch length are proportional to the number of substitutions / site.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">treeRatchet</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parsimony.html">acctran</a></span><span class="op">(</span><span class="va">treeRatchet</span>, <span class="va">primates</span><span class="op">)</span></code></pre></div>
<p>The parsimony ratchet implicitly performs a bootstrap analysis (step 1). We make use of this and store the trees which where visited. This allows us to add bootstrap support values to the tree.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotBS.html">plotBS</a></span><span class="op">(</span><span class="fu"><a href="../reference/midpoint.html">midpoint</a></span><span class="op">(</span><span class="va">treeRatchet</span><span class="op">)</span>, type<span class="op">=</span><span class="st">"phylogram"</span><span class="op">)</span></code></pre></div>
<p><img src="Trees_files/figure-html/midpoint-1.png" width="700"></p>
<div class="section level3">
<h3 id="branch-and-bound">Branch and bound<a class="anchor" aria-label="anchor" href="#branch-and-bound"></a>
</h3>
<p>For data sets with few species it is also possible to find all most parsimonious trees using a branch and bound algorithm <span class="citation">(Hendy and Penny 1982)</span>. For data sets with more than 10 taxa this can take a long time and depends strongly on how tree like the data are. And for more than 20-30 taxa this will take almost forever.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="op">(</span><span class="va">trees</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bab.html">bab</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">primates</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## 1 phylogenetic tree</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="maximum-likelihood">Maximum likelihood<a class="anchor" aria-label="anchor" href="#maximum-likelihood"></a>
</h2>
<p>The last method we will describe in this vignette is Maximum Likelihood (ML) as introduced by Felsenstein <span class="citation">(Felsenstein 1981)</span>. We can easily compute the likelihood for a tree given the data</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">pml</a></span><span class="op">(</span><span class="va">treeNJ</span>, data<span class="op">=</span><span class="va">primates</span><span class="op">)</span>
<span class="va">fit</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">## model: JC </span>
<span class="co">## </span>
<span class="co">## loglikelihood: -3074.952 </span>
<span class="co">## </span>
<span class="co">## unconstrained loglikelihood: -1230.335 </span>
<span class="co">## </span>
<span class="co">## Rate matrix:</span>
<span class="co">##   a c g t</span>
<span class="co">## a 0 1 1 1</span>
<span class="co">## c 1 0 1 1</span>
<span class="co">## g 1 1 0 1</span>
<span class="co">## t 1 1 1 0</span>
<span class="co">## </span>
<span class="co">## Base frequencies:  </span>
<span class="co">## 0.25 0.25 0.25 0.25</span></code></pre>
<p>The function <code>pml</code> returns an object of class <code>pml</code>. This object contains the data, the tree and many different parameters of the model like the likelihood. There are many generic functions for the class <code>pml</code> available, which allow the handling of these objects.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/methods.html" class="external-link">methods</a></span><span class="op">(</span>class<span class="op">=</span><span class="st">"pml"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] AICc   anova  BIC    logLik plot   print  simSeq update vcov  </span>
<span class="co">## see '?methods' for accessing help and source code</span></code></pre>
<p>The object fit just estimated the likelihood for the tree it got supplied, but the branch length are not optimized for the Jukes-Cantor model yet, which can be done with the function <code>optim.pml</code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fitJC</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">optim.pml</a></span><span class="op">(</span><span class="va">fit</span>, rearrangement<span class="op">=</span><span class="st">"NNI"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## optimize edge weights:  -3074.952 --&gt; -3068.417 </span>
<span class="co">## optimize edge weights:  -3068.417 --&gt; -3068.417 </span>
<span class="co">##  optimize topology:  -3068.417 --&gt; -3068.295 </span>
<span class="co">##  NNI moves:  1 </span>
<span class="co">##  optimize topology:  -3068.295 --&gt; -3068.295 </span>
<span class="co">##  NNI moves:  0 </span>
<span class="co">## NNI moves:  1 </span>
<span class="co">## optimize edge weights:  -3068.295 --&gt; -3068.295 </span>
<span class="co">##  optimize topology:  -3068.295 --&gt; -3068.295 </span>
<span class="co">##  NNI moves:  0 </span>
<span class="co">## NNI moves:  0</span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/logLik.html" class="external-link">logLik</a></span><span class="op">(</span><span class="va">fitJC</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## 'log Lik.' -3068.295 (df=25)</span></code></pre>
<p>With the default values<code>pml</code> will estimate a Jukes-Cantor model. The generic function <code>update</code> allows to change parameters. We will change the model to the GTR + <span class="math inline">\(\Gamma(4)\)</span> + I model and then optimize all the parameters.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fitGTR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/update.html" class="external-link">update</a></span><span class="op">(</span><span class="va">fit</span>, k<span class="op">=</span><span class="fl">4</span>, inv<span class="op">=</span><span class="fl">0.2</span><span class="op">)</span>
<span class="va">fitGTR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">optim.pml</a></span><span class="op">(</span><span class="va">fitGTR</span>, model<span class="op">=</span><span class="st">"GTR"</span>, optInv<span class="op">=</span><span class="cn">TRUE</span>, optGamma<span class="op">=</span><span class="cn">TRUE</span>,
    rearrangement <span class="op">=</span> <span class="st">"NNI"</span>, control <span class="op">=</span> <span class="fu"><a href="../reference/pml.control.html">pml.control</a></span><span class="op">(</span>trace <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span>
<span class="va">fitGTR</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">## model: GTR+G(4)+I </span>
<span class="co">## </span>
<span class="co">## loglikelihood: -2608.716 </span>
<span class="co">## </span>
<span class="co">## unconstrained loglikelihood: -1230.335 </span>
<span class="co">## Proportion of invariant sites: 0.005672546 </span>
<span class="co">## Discrete gamma model</span>
<span class="co">## Number of rate categories: 4 </span>
<span class="co">## Shape parameter: 2.987816 </span>
<span class="co">## </span>
<span class="co">## Rate matrix:</span>
<span class="co">##            a           c           g          t</span>
<span class="co">## a  0.0000000  0.61621416 36.10557798  0.3768079</span>
<span class="co">## c  0.6162142  0.00000000  0.00616504 15.1584070</span>
<span class="co">## g 36.1055780  0.00616504  0.00000000  1.0000000</span>
<span class="co">## t  0.3768079 15.15840700  1.00000000  0.0000000</span>
<span class="co">## </span>
<span class="co">## Base frequencies:  </span>
<span class="co">## 0.3934052 0.3791051 0.04024609 0.1872437</span></code></pre>
<p>With the control parameters the thresholds for the fitting process can be changed. Here we want just to suppress output during the fitting process. For larger trees the NNI rearrangements often get stuck in a local maximum. We added two stochastic algorithms to improve topology search. The first (set <code>rearrangement="stochastic"</code>) performs stochastic rearrangements similar as in <span class="citation">(Nguyen et al. 2015)</span>, which makes random NNI permutation to the tree, which than gets optimized to escape local optima. The second option (rearrangement=“ratchet”) perform the likelihood ratchet <span class="citation">(Vos 2003)</span>.</p>
<p>While these algorithms may find better trees they will also take more time.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fitGTR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">optim.pml</a></span><span class="op">(</span><span class="va">fitGTR</span>, model<span class="op">=</span><span class="st">"GTR"</span>, optInv<span class="op">=</span><span class="cn">TRUE</span>, optGamma<span class="op">=</span><span class="cn">TRUE</span>,
    rearrangement <span class="op">=</span> <span class="st">"stochastic"</span>, control <span class="op">=</span> <span class="fu"><a href="../reference/pml.control.html">pml.control</a></span><span class="op">(</span>trace <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span>
<span class="va">fitGTR</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">## model: GTR+G(4)+I </span>
<span class="co">## </span>
<span class="co">## loglikelihood: -2606.75 </span>
<span class="co">## </span>
<span class="co">## unconstrained loglikelihood: -1230.335 </span>
<span class="co">## Proportion of invariant sites: 0.00562857 </span>
<span class="co">## Discrete gamma model</span>
<span class="co">## Number of rate categories: 4 </span>
<span class="co">## Shape parameter: 2.701215 </span>
<span class="co">## </span>
<span class="co">## Rate matrix:</span>
<span class="co">##            a            c            g         t</span>
<span class="co">## a  0.0000000  0.809993516 75.174936272  0.648888</span>
<span class="co">## c  0.8099935  0.000000000  0.004647498 29.027548</span>
<span class="co">## g 75.1749363  0.004647498  0.000000000  1.000000</span>
<span class="co">## t  0.6488880 29.027547743  1.000000000  0.000000</span>
<span class="co">## </span>
<span class="co">## Base frequencies:  </span>
<span class="co">## 0.3963754 0.378828 0.04027169 0.1845249</span></code></pre>
<div class="section level3">
<h3 id="model-selection">Model selection<a class="anchor" aria-label="anchor" href="#model-selection"></a>
</h3>
<p>We can compare nested models for the JC and GTR + <span class="math inline">\(\Gamma(4)\)</span> + I model using likelihood ratio statistic</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/anova.html" class="external-link">anova</a></span><span class="op">(</span><span class="va">fitJC</span>, <span class="va">fitGTR</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Likelihood Ratio Test Table</span>
<span class="co">##   Log lik. Df Df change Diff log lik. Pr(&gt;|Chi|)    </span>
<span class="co">## 1  -3068.3 25                                       </span>
<span class="co">## 2  -2606.8 35        10        923.09  &lt; 2.2e-16 ***</span>
<span class="co">## ---</span>
<span class="co">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></code></pre>
<p>with the Shimodaira-Hasegawa  test</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/SH.test.html">SH.test</a></span><span class="op">(</span><span class="va">fitGTR</span>, <span class="va">fitJC</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##      Trees      ln L Diff ln L p-value</span>
<span class="co">## [1,]     1 -2606.750    0.0000  0.5007</span>
<span class="co">## [2,]     2 -3068.295  461.5449  0.0000</span></code></pre>
<p>or with the AIC</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/AIC.html" class="external-link">AIC</a></span><span class="op">(</span><span class="va">fitJC</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 6186.59</span></code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/AIC.html" class="external-link">AIC</a></span><span class="op">(</span><span class="va">fitGTR</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 5283.5</span></code></pre>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/modelTest.html">AICc</a></span><span class="op">(</span><span class="va">fitGTR</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 5296.358</span></code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/AIC.html" class="external-link">BIC</a></span><span class="op">(</span><span class="va">fitGTR</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 5404.136</span></code></pre>
<p>An alternative is to use the function <code>modelTest</code> to compare different nucleotide or protein models the AIC, AICc or BIC, similar to popular programs ModelTest and ProtTest <span class="citation">(D. Posada and Crandall 1998)</span>, <span class="citation">(David Posada 2008)</span>, <span class="citation">(Abascal, Zardoya, and Posada 2005)</span>. By default available nucleotide or amino acid models available are compared.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/modelTest.html">modelTest</a></span><span class="op">(</span><span class="va">primates</span><span class="op">)</span></code></pre></div>
<p>Here we can select only some common models:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/modelTest.html">modelTest</a></span><span class="op">(</span><span class="va">primates</span>, model<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"JC"</span>, <span class="st">"F81"</span>, <span class="st">"K80"</span>, <span class="st">"HKY"</span>, <span class="st">"SYM"</span>, <span class="st">"GTR"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The results of <code>modelTest</code> is illustrated in following table:</p>
<table class="table">
<thead><tr class="header">
<th align="left">Model</th>
<th align="right">df</th>
<th align="right">logLik</th>
<th align="right">AIC</th>
<th align="right">AICw</th>
<th align="right">AICc</th>
<th align="right">AICcw</th>
<th align="right">BIC</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">JC</td>
<td align="right">25</td>
<td align="right">-3068.42</td>
<td align="right">6186.83</td>
<td align="right">0.00</td>
<td align="right">6193.15</td>
<td align="right">0.00</td>
<td align="right">6273.00</td>
</tr>
<tr class="even">
<td align="left">JC+I</td>
<td align="right">26</td>
<td align="right">-3062.63</td>
<td align="right">6177.26</td>
<td align="right">0.00</td>
<td align="right">6184.10</td>
<td align="right">0.00</td>
<td align="right">6266.87</td>
</tr>
<tr class="odd">
<td align="left">JC+G</td>
<td align="right">26</td>
<td align="right">-3066.92</td>
<td align="right">6185.83</td>
<td align="right">0.00</td>
<td align="right">6192.68</td>
<td align="right">0.00</td>
<td align="right">6275.45</td>
</tr>
<tr class="even">
<td align="left">JC+G+I</td>
<td align="right">27</td>
<td align="right">-3062.63</td>
<td align="right">6179.27</td>
<td align="right">0.00</td>
<td align="right">6186.68</td>
<td align="right">0.00</td>
<td align="right">6272.33</td>
</tr>
<tr class="odd">
<td align="left">F81</td>
<td align="right">28</td>
<td align="right">-2918.17</td>
<td align="right">5892.33</td>
<td align="right">0.00</td>
<td align="right">5900.33</td>
<td align="right">0.00</td>
<td align="right">5988.84</td>
</tr>
<tr class="even">
<td align="left">F81+I</td>
<td align="right">29</td>
<td align="right">-2909.12</td>
<td align="right">5876.24</td>
<td align="right">0.00</td>
<td align="right">5884.85</td>
<td align="right">0.00</td>
<td align="right">5976.20</td>
</tr>
<tr class="odd">
<td align="left">F81+G</td>
<td align="right">29</td>
<td align="right">-2912.56</td>
<td align="right">5883.12</td>
<td align="right">0.00</td>
<td align="right">5891.73</td>
<td align="right">0.00</td>
<td align="right">5983.07</td>
</tr>
<tr class="even">
<td align="left">F81+G+I</td>
<td align="right">30</td>
<td align="right">-2908.52</td>
<td align="right">5877.04</td>
<td align="right">0.00</td>
<td align="right">5886.29</td>
<td align="right">0.00</td>
<td align="right">5980.44</td>
</tr>
<tr class="odd">
<td align="left">K80</td>
<td align="right">26</td>
<td align="right">-2952.94</td>
<td align="right">5957.89</td>
<td align="right">0.00</td>
<td align="right">5964.73</td>
<td align="right">0.00</td>
<td align="right">6047.50</td>
</tr>
<tr class="even">
<td align="left">K80+I</td>
<td align="right">27</td>
<td align="right">-2944.51</td>
<td align="right">5943.02</td>
<td align="right">0.00</td>
<td align="right">5950.43</td>
<td align="right">0.00</td>
<td align="right">6036.08</td>
</tr>
<tr class="odd">
<td align="left">K80+G</td>
<td align="right">27</td>
<td align="right">-2944.76</td>
<td align="right">5943.53</td>
<td align="right">0.00</td>
<td align="right">5950.94</td>
<td align="right">0.00</td>
<td align="right">6036.59</td>
</tr>
<tr class="even">
<td align="left">K80+G+I</td>
<td align="right">28</td>
<td align="right">-2942.34</td>
<td align="right">5940.68</td>
<td align="right">0.00</td>
<td align="right">5948.68</td>
<td align="right">0.00</td>
<td align="right">6037.19</td>
</tr>
<tr class="odd">
<td align="left">HKY</td>
<td align="right">29</td>
<td align="right">-2625.03</td>
<td align="right">5308.07</td>
<td align="right">0.00</td>
<td align="right">5316.68</td>
<td align="right">0.00</td>
<td align="right">5408.02</td>
</tr>
<tr class="even">
<td align="left">HKY+I</td>
<td align="right">30</td>
<td align="right">-2621.27</td>
<td align="right">5302.54</td>
<td align="right">0.00</td>
<td align="right">5311.79</td>
<td align="right">0.00</td>
<td align="right">5405.94</td>
</tr>
<tr class="odd">
<td align="left">HKY+G</td>
<td align="right">30</td>
<td align="right">-2612.64</td>
<td align="right">5285.28</td>
<td align="right">0.18</td>
<td align="right">5294.54</td>
<td align="right">0.45</td>
<td align="right">5388.68</td>
</tr>
<tr class="even">
<td align="left">HKY+G+I</td>
<td align="right">31</td>
<td align="right">-2612.45</td>
<td align="right">5286.89</td>
<td align="right">0.08</td>
<td align="right">5296.81</td>
<td align="right">0.14</td>
<td align="right">5393.74</td>
</tr>
<tr class="odd">
<td align="left">SYM</td>
<td align="right">30</td>
<td align="right">-2813.90</td>
<td align="right">5687.79</td>
<td align="right">0.00</td>
<td align="right">5697.05</td>
<td align="right">0.00</td>
<td align="right">5791.19</td>
</tr>
<tr class="even">
<td align="left">SYM+I</td>
<td align="right">31</td>
<td align="right">-2811.73</td>
<td align="right">5685.46</td>
<td align="right">0.00</td>
<td align="right">5695.38</td>
<td align="right">0.00</td>
<td align="right">5792.31</td>
</tr>
<tr class="odd">
<td align="left">SYM+G</td>
<td align="right">31</td>
<td align="right">-2804.68</td>
<td align="right">5671.36</td>
<td align="right">0.00</td>
<td align="right">5681.28</td>
<td align="right">0.00</td>
<td align="right">5778.20</td>
</tr>
<tr class="even">
<td align="left">SYM+G+I</td>
<td align="right">32</td>
<td align="right">-2804.67</td>
<td align="right">5673.34</td>
<td align="right">0.00</td>
<td align="right">5683.95</td>
<td align="right">0.00</td>
<td align="right">5783.63</td>
</tr>
<tr class="odd">
<td align="left">GTR</td>
<td align="right">33</td>
<td align="right">-2618.64</td>
<td align="right">5303.28</td>
<td align="right">0.00</td>
<td align="right">5314.61</td>
<td align="right">0.00</td>
<td align="right">5417.02</td>
</tr>
<tr class="even">
<td align="left">GTR+I</td>
<td align="right">34</td>
<td align="right">-2613.58</td>
<td align="right">5295.17</td>
<td align="right">0.00</td>
<td align="right">5307.25</td>
<td align="right">0.00</td>
<td align="right">5412.36</td>
</tr>
<tr class="odd">
<td align="left">GTR+G</td>
<td align="right">34</td>
<td align="right">-2607.67</td>
<td align="right">5283.33</td>
<td align="right">0.47</td>
<td align="right">5295.41</td>
<td align="right">0.29</td>
<td align="right">5400.52</td>
</tr>
<tr class="even">
<td align="left">GTR+G+I</td>
<td align="right">35</td>
<td align="right">-2607.21</td>
<td align="right">5284.43</td>
<td align="right">0.27</td>
<td align="right">5297.29</td>
<td align="right">0.11</td>
<td align="right">5405.07</td>
</tr>
</tbody>
</table>
<p>The thresholds for the optimization in <code>modelTest</code> are not as strict as for <code>optim.pml</code> and no tree rearrangements are performed. As <code>modelTest</code> computes and optimizes a lot of models it would be a waste of computer time not to save these results. The results are saved as call together with the optimized trees in an environment and the function <code>as.pml</code> evaluates this call to get a <code>pml</code> object back to use for further optimization or analysis.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="op">(</span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">as.pml</a></span><span class="op">(</span><span class="va">mt</span>, <span class="st">"HKY+G+I"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">## model: HKY+G(4)+I </span>
<span class="co">## </span>
<span class="co">## loglikelihood: -2612.446 </span>
<span class="co">## </span>
<span class="co">## unconstrained loglikelihood: -1230.335 </span>
<span class="co">## Proportion of invariant sites: 0.002733786 </span>
<span class="co">## Discrete gamma model</span>
<span class="co">## Number of rate categories: 4 </span>
<span class="co">## Shape parameter: 2.127185 </span>
<span class="co">## </span>
<span class="co">## Rate matrix:</span>
<span class="co">##          a        c        g        t</span>
<span class="co">## a  0.00000  1.00000 56.03063  1.00000</span>
<span class="co">## c  1.00000  0.00000  1.00000 56.03063</span>
<span class="co">## g 56.03063  1.00000  0.00000  1.00000</span>
<span class="co">## t  1.00000 56.03063  1.00000  0.00000</span>
<span class="co">## </span>
<span class="co">## Base frequencies:  </span>
<span class="co">## 0.4204974 0.3622544 0.04393528 0.1733129</span></code></pre>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="op">(</span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">as.pml</a></span><span class="op">(</span><span class="va">mt</span>, <span class="st">"BIC"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">## model: HKY+G(4) </span>
<span class="co">## </span>
<span class="co">## loglikelihood: -2612.641 </span>
<span class="co">## </span>
<span class="co">## unconstrained loglikelihood: -1230.335 </span>
<span class="co">## Discrete gamma model</span>
<span class="co">## Number of rate categories: 4 </span>
<span class="co">## Shape parameter: 2.145192 </span>
<span class="co">## </span>
<span class="co">## Rate matrix:</span>
<span class="co">##          a        c        g        t</span>
<span class="co">## a  0.00000  1.00000 51.43709  1.00000</span>
<span class="co">## c  1.00000  0.00000  1.00000 51.43709</span>
<span class="co">## g 51.43709  1.00000  0.00000  1.00000</span>
<span class="co">## t  1.00000 51.43709  1.00000  0.00000</span>
<span class="co">## </span>
<span class="co">## Base frequencies:  </span>
<span class="co">## 0.4193408 0.3628885 0.04388863 0.173882</span></code></pre>
</div>
<div class="section level3">
<h3 id="bootstrap-1">Bootstrap<a class="anchor" aria-label="anchor" href="#bootstrap-1"></a>
</h3>
<p>At last we may want to apply bootstrap to test how well the edges of the tree are supported:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootstrap.pml.html">bootstrap.pml</a></span><span class="op">(</span><span class="va">fitJC</span>, bs<span class="op">=</span><span class="fl">100</span>, optNni<span class="op">=</span><span class="cn">TRUE</span>,
    control <span class="op">=</span> <span class="fu"><a href="../reference/pml.control.html">pml.control</a></span><span class="op">(</span>trace <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Now we can plot the tree with the bootstrap support values on the edges and also look at <code>consensusNet</code> to identify potential conflict.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotBS.html">plotBS</a></span><span class="op">(</span><span class="fu"><a href="../reference/midpoint.html">midpoint</a></span><span class="op">(</span><span class="va">fitJC</span><span class="op">$</span><span class="va">tree</span><span class="op">)</span>, <span class="va">bs</span>, p <span class="op">=</span> <span class="fl">50</span>, type<span class="op">=</span><span class="st">"p"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="Trees_files/figure-html/plotBS-1.png" alt="Tree with bootstrap support. Unrooted tree (midpoint rooted) with bootstrap support values." width="700"><p class="caption">
Tree with bootstrap support. Unrooted tree (midpoint rooted) with bootstrap support values.
</p>
</div>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cnet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/consensusNet.html">consensusNet</a></span><span class="op">(</span><span class="va">bs</span>, p<span class="op">=</span><span class="fl">0.2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cnet</span>, show.edge.label<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="Trees_files/figure-html/ConsensusNet-1.png" alt="ConsensusNet from the bootstrap sample." width="700"><p class="caption">
ConsensusNet from the bootstrap sample.
</p>
</div>
</div>
<div class="section level3">
<h3 id="molecular-dating-with-a-strict-clock">Molecular dating with a strict clock<a class="anchor" aria-label="anchor" href="#molecular-dating-with-a-strict-clock"></a>
</h3>
<p>We implemented a strict clock as described in <span class="citation">(Felsenstein 2004)</span>, p. ???.<br>
So far the starting tree needs to be ultrametric, so we use the UPGMA tree. When we optimize the tree with <code>optim.pml</code> we have to make sure set we set <code>optRooted = TRUE</code>!</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fit_strict</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">pml</a></span><span class="op">(</span><span class="va">treeUPGMA</span>, data<span class="op">=</span><span class="va">primates</span>, k<span class="op">=</span><span class="fl">4</span>, bf<span class="op">=</span><span class="fu"><a href="../reference/baseFreq.html">baseFreq</a></span><span class="op">(</span><span class="va">primates</span><span class="op">)</span><span class="op">)</span>
<span class="va">fit_strict</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">optim.pml</a></span><span class="op">(</span><span class="va">fit_strict</span>, model<span class="op">=</span><span class="st">"GTR"</span>, optRooted <span class="op">=</span> <span class="cn">TRUE</span>, 
                        rearrangement <span class="op">=</span> <span class="st">"NNI"</span>, optGamma <span class="op">=</span> <span class="cn">TRUE</span>, optInv <span class="op">=</span> <span class="cn">TRUE</span>, 
                        control <span class="op">=</span> <span class="fu"><a href="../reference/pml.control.html">pml.control</a></span><span class="op">(</span>trace <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fit_strict</span><span class="op">)</span></code></pre></div>
<p><img src="Trees_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>Several analyses, e.g.<code>bootstrap</code> and <code>modelTest</code>, can be computationally demanding, but as nowadays most computers have several cores one can distribute the computations using the <em>parallel</em> package. However it is only possible to use this approach if R is running from command line (“X11”), but not using a GUI (for example “Aqua” on Macs) and unfortunately the <em>parallel</em> package does not work at all under Windows. </p>
</div>
</div>
<div class="section level2">
<h2 id="appendix">Appendix<a class="anchor" aria-label="anchor" href="#appendix"></a>
</h2>
<div class="section level3">
<h3 id="standard-scripts-for-nucleotide-analysis">Standard scripts for nucleotide analysis<a class="anchor" aria-label="anchor" href="#standard-scripts-for-nucleotide-analysis"></a>
</h3>
<p>Here we provide two standard scripts which can be adapted for the most common tasks. Most likely the arguments for <code>read.phyDat</code> have to be adapted to accommodate your file format. Both scripts assume that the <em>parallel</em> package works on your platform, see comments above.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/KlausVigo/phangorn" class="external-link">phangorn</a></span><span class="op">)</span>
<span class="va">file</span> <span class="op">&lt;-</span> <span class="st">"myfile"</span>
<span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read.phyDat.html">read.phyDat</a></span><span class="op">(</span><span class="va">file</span><span class="op">)</span>
<span class="va">dm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dist.hamming.html">dist.ml</a></span><span class="op">(</span><span class="va">dat</span>, <span class="st">"F81"</span><span class="op">)</span>
<span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/NJ.html">NJ</a></span><span class="op">(</span><span class="va">dm</span><span class="op">)</span>
<span class="co"># as alternative for a starting tree:</span>
<span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parsimony.html">pratchet</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span>          <span class="co"># parsimony tree</span>
<span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/designTree.html">nnls.phylo</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">dm</span><span class="op">)</span>   <span class="co"># need edge weights</span>


<span class="co"># 1. alternative: quick and dirty: GTR + G</span>
<span class="va">fitStart</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">pml</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">dat</span>, k<span class="op">=</span><span class="fl">4</span><span class="op">)</span>
<span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">optim.pml</a></span><span class="op">(</span><span class="va">fitStart</span>, model<span class="op">=</span><span class="st">"GTR"</span>, optGamma<span class="op">=</span><span class="cn">TRUE</span>, rearrangement<span class="op">=</span><span class="st">"stochastic"</span><span class="op">)</span>

<span class="co"># 2. alternative: prepare with modelTest</span>
<span class="va">mt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/modelTest.html">modelTest</a></span><span class="op">(</span><span class="va">dat</span>, tree<span class="op">=</span><span class="va">tree</span>, multicore<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="va">mt</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">mt</span><span class="op">$</span><span class="va">AICc</span><span class="op">)</span>,<span class="op">]</span>
<span class="co"># choose best model from the table according to AICc</span>
<span class="va">fitStart</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">as.pml</a></span><span class="op">(</span><span class="va">mt</span>, <span class="st">"AICc"</span><span class="op">)</span>

<span class="co"># equivalent to:   fitStart = eval(get("GTR+G+I", env), env)</span>
<span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">optim.pml</a></span><span class="op">(</span><span class="va">fitStart</span>, rearrangement <span class="op">=</span> <span class="st">"stochastic"</span>,
    optGamma<span class="op">=</span><span class="cn">TRUE</span>, optInv<span class="op">=</span><span class="cn">TRUE</span>, model<span class="op">=</span><span class="st">"GTR"</span><span class="op">)</span>
<span class="va">bs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootstrap.pml.html">bootstrap.pml</a></span><span class="op">(</span><span class="va">fit</span>, bs<span class="op">=</span><span class="fl">100</span>, optNni<span class="op">=</span><span class="cn">TRUE</span>, multicore<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div style="page-break-after: always;"></div>
</div>
<div class="section level3">
<h3 id="standard-scripts-for-amino-acid-analysis">Standard scripts for amino acid analysis<a class="anchor" aria-label="anchor" href="#standard-scripts-for-amino-acid-analysis"></a>
</h3>
<p>You can specify different several models build in which you can specify, e.g. “WAG”, “JTT”, “Dayhoff”, “LG”. Optimizing the rate matrix for amino acids is possible, but would take a long, a very long time and you will need to have a large alignment to estimate all the parameters. So make sure to set <code>optBf=FALSE</code> and <code>optQ=FALSE</code> in the function <code>optim.pml</code>, which is also the default.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/KlausVigo/phangorn" class="external-link">phangorn</a></span><span class="op">)</span>
<span class="va">file</span> <span class="op">&lt;-</span> <span class="st">"myfile"</span>
<span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read.phyDat.html">read.phyDat</a></span><span class="op">(</span><span class="va">file</span>, type <span class="op">=</span> <span class="st">"AA"</span><span class="op">)</span>
<span class="va">dm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dist.hamming.html">dist.ml</a></span><span class="op">(</span><span class="va">dat</span>, model<span class="op">=</span><span class="st">"JTT"</span><span class="op">)</span>
<span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/NJ.html">NJ</a></span><span class="op">(</span><span class="va">dm</span><span class="op">)</span>

<span class="co"># parallel will only work safely from command line</span>
<span class="co"># and not at all windows</span>
<span class="op">(</span><span class="va">mt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/modelTest.html">modelTest</a></span><span class="op">(</span><span class="va">dat</span>, model<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"JTT"</span>, <span class="st">"LG"</span>, <span class="st">"WAG"</span><span class="op">)</span>,
    multicore<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="co"># run all available amino acid models</span>
<span class="op">(</span><span class="va">mt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/modelTest.html">modelTest</a></span><span class="op">(</span><span class="va">dat</span>, model<span class="op">=</span><span class="st">"all"</span>, multicore<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>

<span class="va">fitStart</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">as.pml</a></span><span class="op">(</span><span class="va">mt</span><span class="op">)</span>

<span class="va">fitNJ</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">pml</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">dat</span>, model<span class="op">=</span><span class="st">"JTT"</span>, k<span class="op">=</span><span class="fl">4</span>, inv<span class="op">=</span><span class="fl">.2</span><span class="op">)</span>
<span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">optim.pml</a></span><span class="op">(</span><span class="va">fitNJ</span>, rearrangement <span class="op">=</span> <span class="st">"stochastic"</span>,
    optInv<span class="op">=</span><span class="cn">TRUE</span>, optGamma<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="va">fit</span>

<span class="va">bs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootstrap.pml.html">bootstrap.pml</a></span><span class="op">(</span><span class="va">fit</span>, bs<span class="op">=</span><span class="fl">100</span>, optNni<span class="op">=</span><span class="cn">TRUE</span>, multicore<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span class="co">## R version 4.2.0 (2022-04-22)</span>
<span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">## Running under: Ubuntu 20.04.4 LTS</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0</span>
<span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span>
<span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span>
<span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span>
<span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">## [1] knitr_1.39          phangorn_2.9.0.9000 ape_5.6-2          </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##  [1] Rcpp_1.0.8.3      highr_0.9         bslib_0.3.1       compiler_4.2.0   </span>
<span class="co">##  [5] jquerylib_0.1.4   prettydoc_0.4.1   tools_4.2.0       digest_0.6.29    </span>
<span class="co">##  [9] jsonlite_1.8.0    evaluate_0.15     memoise_2.0.1     nlme_3.1-157     </span>
<span class="co">## [13] lattice_0.20-45   pkgconfig_2.0.3   rlang_1.0.2       Matrix_1.4-1     </span>
<span class="co">## [17] fastmatch_1.1-3   igraph_1.3.2      cli_3.3.0         yaml_2.3.5       </span>
<span class="co">## [21] parallel_4.2.0    pkgdown_2.0.4     xfun_0.31         fastmap_1.1.0    </span>
<span class="co">## [25] stringr_1.4.0     desc_1.4.1        generics_0.1.2    fs_1.5.2         </span>
<span class="co">## [29] sass_0.4.1        systemfonts_1.0.4 rprojroot_2.0.3   grid_4.2.0       </span>
<span class="co">## [33] R6_2.5.1          textshaping_0.3.6 rmarkdown_2.14    purrr_0.3.4      </span>
<span class="co">## [37] magrittr_2.0.3    codetools_0.2-18  htmltools_0.5.2   quadprog_1.5-8   </span>
<span class="co">## [41] ragg_1.2.2        stringi_1.7.6     cachem_1.0.6</span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Abascal2005" class="csl-entry">
Abascal, Federico, Rafael Zardoya, and David Posada. 2005. <span>“ProtTest: Selection of Best-Fit Models of Protein Evolution.”</span> <em>Bioinformatics</em> 21 (9): 2104–5. <a href="https://doi.org/10.1093/bioinformatics/bti263" class="external-link">https://doi.org/10.1093/bioinformatics/bti263</a>.
</div>
<div id="ref-Felsenstein1981" class="csl-entry">
Felsenstein, Joseph. 1981. <span>“Evolutionary Trees from DNA Sequences: A Maxumum Likelihood Approach.”</span> <em>Journal of Molecular Evolution</em> 17: 368–76.
</div>
<div id="ref-Felsenstein2004" class="csl-entry">
———. 2004. <em>Inferring Phylogenies</em>. Sunderland: Sinauer Associates.
</div>
<div id="ref-Hendy1982" class="csl-entry">
Hendy, M. D., and D. Penny. 1982. <span>“Branch and Bound Algorithms to Determine Minimal Evolutionary Trees.”</span> <em>Math. Biosc.</em> 59: 277–90.
</div>
<div id="ref-Nguyen2015" class="csl-entry">
Nguyen, Lam-Tung, Heiko A. Schmidt, Arndt von Haeseler, and Bui Quang Minh. 2015. <span>“IQ-TREE: A Fast and Effective Stochastic Algorithm for Estimating Maximum-Likelihood Phylogenies.”</span> <em>Molecular Biology and Evolution</em> 32 (1): 268–74. <a href="https://doi.org/10.1093/molbev/msu300" class="external-link">https://doi.org/10.1093/molbev/msu300</a>.
</div>
<div id="ref-Nixon1999" class="csl-entry">
Nixon, K. 1999. <span>“The Parsimony Ratchet, a New Method for Rapid Rarsimony Analysis.”</span> <em>Cladistics</em> 15: 407–14.
</div>
<div id="ref-Paradis2012" class="csl-entry">
Paradis, Emmanuel. 2012. <em>Analysis of Phylogenetics and Evolution with r</em>. Second. New York: Springer.
</div>
<div id="ref-Paradis2018" class="csl-entry">
Paradis, Emmanuel, and Klaus Schliep. 2019. <span>“Ape 5.0: An Environment for Modern Phylogenetics and Evolutionary Analyses in r.”</span> <em>Bioinformatics</em> 35 (3): 526–28. <a href="https://doi.org/10.1093/bioinformatics/bty633" class="external-link">https://doi.org/10.1093/bioinformatics/bty633</a>.
</div>
<div id="ref-Posada2008" class="csl-entry">
Posada, David. 2008. <span>“<span class="nocase">jModelTest</span>: Phylogenetic Model Averaging.”</span> <em>Molecular Biology and Evolution</em> 25 (7): 1253–56. <a href="https://doi.org/10.1093/molbev/msn083" class="external-link">https://doi.org/10.1093/molbev/msn083</a>.
</div>
<div id="ref-Posada1998" class="csl-entry">
Posada, D., and K. A. Crandall. 1998. <span>“<span>MODELTEST</span>: Testing the Model of <span>DNA</span> Substitution.”</span> <em>Bioinformatics</em> 14 (9): 817–18.
</div>
<div id="ref-Saitou1987" class="csl-entry">
Saitou, N., and M. Nei. 1987. <span>“The Neighbor-Joining Method - a New Method for Reconstructing Phylogenetic Trees.”</span> <em>Molecular Biology and Evolution</em> 4 (4): 406–25.
</div>
<div id="ref-Schliep2011" class="csl-entry">
Schliep, Klaus Peter. 2011. <span>“Phangorn: Phylogenetic Analysis in <span>R</span>.”</span> <em>Bioinformatics</em> 27 (4): 592–93. <a href="https://doi.org/10.1093/bioinformatics/btq706" class="external-link">https://doi.org/10.1093/bioinformatics/btq706</a>.
</div>
<div id="ref-Studier1988" class="csl-entry">
Studier, J. A., and K. J. Keppler. 1988. <span>“A Note on the Neighbor-Joining Algorithm of Saitou and Nei.”</span> <em>Molecular Biology and Evolution</em> 5 (6): 729–31.
</div>
<div id="ref-Vos2003" class="csl-entry">
Vos, R. A. 2003. <span>“Accelerated Likelihood Surface Exploration: The Likelihood Ratchet.”</span> <em>Systematic Biology</em> 52 (3): 368–73.
</div>
<div id="ref-Yang2006" class="csl-entry">
Yang, Ziheng. 2006. <em>Computational Molecular Evolution</em>. Oxford: Oxford University Press.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Klaus Schliep, Emmanuel Paradis, Leonardo de Oliveira Martins, Alastair Potts, Tim W. White.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.4.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
